pgrx/src/callconv.rs

```diff
@@ -334,9 +334,23 @@ pub unsafe trait RetAbi: Sized {
     unsafe fn finish_call_fcinfo(_fcinfo: pg_sys::FunctionCallInfo) {}
 }
 
-/// A simplified blanket RetAbi
+/// Simplified variant of *RetAbi*.
+///
+/// In order to handle all of the nuances of the calling convention for returning result sets from
+/// Postgres functions, pgrx uses a very complicated trait, RetAbi. In practice, however, most
+/// types do not need to think about its many sharp-edged cases. Instead, they should implement
+/// this simplified trait, BoxRet. A blanket impl of RetAbi for BoxRet takes care of the rest.
 pub unsafe trait BoxRet: Sized {
-    unsafe fn box_in_fcinfo(self, fcinfo: pg_sys::FunctionCallInfo) -> pg_sys::Datum;
+    #[doc(hidden)]
+    unsafe fn box_in_fcinfo(self, fcinfo: pg_sys::FunctionCallInfo) -> pg_sys::Datum {
+        unsafe { Self::box_into(self, &mut FcInfo::from_ptr(fcinfo)).sans_lifetime() }
+    }
+
+    /// # Safety
+    /// You have to actually return the resulting Datum to the function.
+    unsafe fn box_into<'fcx>(self, fcinfo: &mut FcInfo<'fcx>) -> Datum<'fcx> {
+        unsafe { fcinfo.return_raw_datum(self.box_in_fcinfo(fcinfo.0)) }
+    }
```

Comment:
This safety constraint isn't a big concern because generally "user code" should not be calling this function directly, but rather implementing it so that it can _be_ called.